File: mobile/ios/Runner/Upload/TusUploadApiImpl.swift
import Foundation
import Flutter
import TUSKit

class TusUploadApiImpl: NSObject, TusUploadApi, FlutterPlugin {
  static let name = "TusUploadApi"

  private var tusClient: TUSClient?
  private var callbackApi: TusUploadCallbackApi?
  private var activeUploads: [String: UUID] = [:]
  private var uploadIdToDeviceAssetId: [UUID: String] = [:]
  private var currentServerEndpoint: String?

  static func register(with registrar: FlutterPluginRegistrar) {
    let instance = TusUploadApiImpl()
    TusUploadApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    instance.callbackApi = TusUploadCallbackApi(binaryMessenger: registrar.messenger())
    registrar.publish(instance)
  }

  private func getTusClient(serverEndpoint: String) throws -> TUSClient {
    if let client = tusClient, currentServerEndpoint == serverEndpoint {
      return client
    }

    guard let uploadURL = URL(string: "\(serverEndpoint)/upload") else {
      throw NSError(domain: "TusUploadApiImpl", code: 1, userInfo: [NSLocalizedDescriptionKey: "Invalid server endpoint"])
    }

    do {
      let client = try TUSClient(
        server: uploadURL,
        sessionIdentifier: "immich-tus-session",
        sessionConfiguration: .default,
        storageDirectory: FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!.appendingPathComponent("tus-uploads"),
        chunkSize: 50 * 1024 * 1024 // 50MB chunks for better performance
      )

      client.delegate = self
      tusClient = client
      currentServerEndpoint = serverEndpoint
      return client
    } catch {
      tusClient = nil
      currentServerEndpoint = nil
      throw error
    }
  }

  func startUpload(data: TusUploadData, completion: @escaping (Result<String, Error>) -> Void) {
    do {
      let client = try getTusClient(serverEndpoint: data.serverEndpoint)

      let fileURL = URL(fileURLWithPath: data.filePath)

      // Note: Do NOT Base64 encode values here - TUSKit handles encoding automatically
      var metadata: [String: String] = [:]
      metadata["filename"] = data.filename
      metadata["deviceAssetId"] = data.deviceAssetId
      metadata["deviceId"] = data.deviceId
      metadata["fileCreatedAt"] = data.fileCreatedAt
      metadata["fileModifiedAt"] = data.fileModifiedAt
      metadata["isFavorite"] = String(data.isFavorite)

      if let duration = data.duration {
        metadata["duration"] = duration
      }
      if let livePhotoVideoId = data.livePhotoVideoId {
        metadata["livePhotoVideoId"] = livePhotoVideoId
      }

      var headers: [String: String] = [:]
      for (key, value) in data.headers {
        headers[key] = value
      }

      let uploadId = try client.uploadFileAt(
        filePath: fileURL,
        customHeaders: headers,
        context: metadata
      )

      activeUploads[data.deviceAssetId] = uploadId
      uploadIdToDeviceAssetId[uploadId] = data.deviceAssetId

      callbackApi?.onStatusChange(
        update: TusStatusUpdate(
          uploadId: data.deviceAssetId,
          status: "uploading",
          assetId: nil,
          error: nil
        )
      ) { _ in }

      completion(.success(data.deviceAssetId))

    } catch {
      completion(.failure(error))
    }
  }

  func cancelUpload(uploadId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
    guard let tusUploadId = activeUploads[uploadId] else {
      completion(.success(false))
      return
    }

    do {
      try tusClient?.cancel(id: tusUploadId)
      activeUploads.removeValue(forKey: uploadId)
      uploadIdToDeviceAssetId.removeValue(forKey: tusUploadId)
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func cancelAllUploads(completion: @escaping (Result<Void, Error>) -> Void) {
    tusClient?.stopAndCancelAll()
    activeUploads.removeAll()
    uploadIdToDeviceAssetId.removeAll()
    completion(.success(()))
  }

  func getUploadOffset(uploadId: String, completion: @escaping (Result<Int64?, Error>) -> Void) {
    // TUSKit doesn't expose offset directly, return nil
    completion(.success(nil))
  }

  func getPendingUploadIds(completion: @escaping (Result<[String], Error>) -> Void) {
    completion(.success(Array(activeUploads.keys)))
  }

  private func fetchAssetIdFromServer(uploadURL: URL) async -> String? {
    var request = URLRequest(url: uploadURL)
    request.httpMethod = "HEAD"
    request.setValue("1.0.0", forHTTPHeaderField: "Tus-Resumable")

    do {
      let (_, response) = try await URLSession.shared.data(for: request)
      if let httpResponse = response as? HTTPURLResponse {
        return httpResponse.value(forHTTPHeaderField: "X-Immich-Asset-Id")
      }
    } catch {
      NSLog("Failed to fetch asset ID from server: \(error)")
    }

    return nil
  }
}

// MARK: - TUSClientDelegate

extension TusUploadApiImpl: TUSClientDelegate {
  func didStartUpload(id: UUID, context: [String: String]?, client: TUSClient) {
    guard let deviceAssetId = uploadIdToDeviceAssetId[id] else { return }

    callbackApi?.onStatusChange(
      update: TusStatusUpdate(
        uploadId: deviceAssetId,
        status: "uploading",
        assetId: nil,
        error: nil
      )
    ) { _ in }
  }

  func didFinishUpload(id: UUID, url: URL, context: [String: String]?, client: TUSClient) {
    guard let deviceAssetId = uploadIdToDeviceAssetId[id] else { return }

    let assetIdFromContext = context?["x-immich-asset-id"] ?? context?["X-Immich-Asset-Id"]

    if let assetId = assetIdFromContext, !assetId.isEmpty {
      completeUpload(deviceAssetId: deviceAssetId, tusId: id, assetId: assetId)
    } else {
      Task {
        let assetId = await fetchAssetIdFromServer(uploadURL: url)
        completeUpload(deviceAssetId: deviceAssetId, tusId: id, assetId: assetId)
      }
    }
  }

  private func completeUpload(deviceAssetId: String, tusId: UUID, assetId: String?) {
    callbackApi?.onStatusChange(
      update: TusStatusUpdate(
        uploadId: deviceAssetId,
        status: "completed",
        assetId: assetId,
        error: nil
      )
    ) { _ in }

    activeUploads.removeValue(forKey: deviceAssetId)
    uploadIdToDeviceAssetId.removeValue(forKey: tusId)
  }

  func uploadFailed(id: UUID, error: Error, context: [String: String]?, client: TUSClient) {
    guard let deviceAssetId = uploadIdToDeviceAssetId[id] else { return }

    callbackApi?.onStatusChange(
      update: TusStatusUpdate(
        uploadId: deviceAssetId,
        status: "failed",
        assetId: nil,
        error: error.localizedDescription
      )
    ) { _ in }

    activeUploads.removeValue(forKey: deviceAssetId)
    uploadIdToDeviceAssetId.removeValue(forKey: id)
  }

  func fileError(error: TUSClientError, client: TUSClient) {
    NSLog("TUS file error: \(error.localizedDescription)")
  }

  func totalProgress(bytesUploaded: Int, totalBytes: Int, client: TUSClient) {
    // This is global progress, not per-upload
  }

  func progressFor(id: UUID, context: [String: String]?, bytesUploaded: Int, totalBytes: Int, client: TUSClient) {
    guard let deviceAssetId = uploadIdToDeviceAssetId[id] else { return }

    callbackApi?.onProgress(
      update: TusProgressUpdate(
        uploadId: deviceAssetId,
        bytesUploaded: Int64(bytesUploaded),
        totalBytes: Int64(totalBytes)
      )
    ) { _ in }
  }
}

File: mobile/android/app/src/main/kotlin/app/alextran/immich/upload/TusUploadApiImpl.kt
package app.alextran.immich.upload

import android.content.Context
import app.alextran.immich.core.ImmichPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.tus.java.client.TusClient
import io.tus.java.client.TusUpload
import io.tus.java.client.TusUploader
import kotlinx.coroutines.*
import java.io.File
import java.net.URL
import java.util.concurrent.ConcurrentHashMap

class TusUploadApiImpl(context: Context) : ImmichPlugin(), TusUploadApi {
  companion object {
    const val name = "TusUploadApi"
    private const val CHUNK_SIZE = 50 * 1024 * 1024 // 50MB chunks for better performance
  }

  private val ctx: Context = context.applicationContext
  private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
  private val activeUploads = ConcurrentHashMap<String, Job>()
  private var callbackApi: TusUploadCallbackApi? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    super.onAttachedToEngine(binding)
    TusUploadApi.setUp(binding.binaryMessenger, this)
    callbackApi = TusUploadCallbackApi(binding.binaryMessenger)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    super.onDetachedFromEngine(binding)
    TusUploadApi.setUp(binding.binaryMessenger, null)
    cancelAllUploadsSync()
    scope.cancel()
    callbackApi = null
  }

  override fun startUpload(data: TusUploadData, callback: (Result<String>) -> Unit) {
    val uploadId = data.deviceAssetId

    val job = scope.launch {
      try {
        val file = File(data.filePath)
        if (!file.exists()) {
          callback(Result.failure(Exception("File not found: ${data.filePath}")))
          return@launch
        }

        val client = TusClient().apply {
          uploadCreationURL = URL("${data.serverEndpoint}/upload")
          enableResuming(TusPreferencesURLStore(ctx, uploadId))
        }

        // Note: Do NOT set Tus-Resumable here - tus-java-client sets it automatically
        // Setting it twice results in "1.0.0, 1.0.0" which fails validation
        client.headers = data.headers

        val upload = TusUpload(file).apply {
          // Set metadata
          metadata = buildMetadata(data)
        }

        withContext(Dispatchers.Main) {
          callbackApi?.onStatusChange(
            TusStatusUpdate(
              uploadId = uploadId,
              status = "uploading",
              assetId = null,
              error = null
            )
          ) {}
        }

        val uploader = client.resumeOrCreateUpload(upload)
        uploader.chunkSize = CHUNK_SIZE

        var bytesUploaded = uploader.offset
        val totalBytes = upload.size

        while (uploader.uploadChunk() > -1) {
          val job = activeUploads[uploadId]
          if (job == null || !job.isActive) {
            uploader.finish()
            throw CancellationException("Upload cancelled")
          }

          bytesUploaded = uploader.offset

          withContext(Dispatchers.Main) {
            callbackApi?.onProgress(
              TusProgressUpdate(
                uploadId = uploadId,
                bytesUploaded = bytesUploaded,
                totalBytes = totalBytes
              )
            ) {}
          }
        }

        uploader.finish()

        // Fetch asset ID from server via HEAD request using the upload URL from uploader
        val assetId = fetchAssetIdFromServer(uploader.uploadURL, data.headers)

        val urlStore = TusPreferencesURLStore(ctx, uploadId)
        urlStore.clear()

        withContext(Dispatchers.Main) {
          callbackApi?.onStatusChange(
            TusStatusUpdate(
              uploadId = uploadId,
              status = "completed",
              assetId = assetId,
              error = null
            )
          ) {}
        }

        callback(Result.success(uploadId))

      } catch (e: CancellationException) {
        callback(Result.failure(Exception("Upload cancelled")))
        withContext(Dispatchers.Main) {
          callbackApi?.onStatusChange(
            TusStatusUpdate(
              uploadId = uploadId,
              status = "cancelled",
              assetId = null,
              error = "Upload cancelled"
            )
          ) {}
        }
      } catch (e: Exception) {
        callback(Result.failure(e))
        withContext(Dispatchers.Main) {
          callbackApi?.onStatusChange(
            TusStatusUpdate(
              uploadId = uploadId,
              status = "failed",
              assetId = null,
              error = e.message
            )
          ) {}
        }
      } finally {
        activeUploads.remove(uploadId)
      }
    }

    activeUploads[uploadId] = job
  }

  override fun cancelUpload(uploadId: String, callback: (Result<Boolean>) -> Unit) {
    val job = activeUploads[uploadId]
    if (job != null) {
      job.cancel()
      activeUploads.remove(uploadId)
      callback(Result.success(true))
    } else {
      callback(Result.success(false))
    }
  }

  override fun cancelAllUploads(callback: (Result<Unit>) -> Unit) {
    cancelAllUploadsSync()
    callback(Result.success(Unit))
  }

  private fun cancelAllUploadsSync() {
    activeUploads.values.forEach { it.cancel() }
    activeUploads.clear()
  }

  override fun getUploadOffset(uploadId: String, callback: (Result<Long?>) -> Unit) {
    scope.launch {
      try {
        val urlStore = TusPreferencesURLStore(ctx, uploadId)
        val url = urlStore.get(uploadId)
        if (url != null) {
          // We would need to make a HEAD request to get the offset
          callback(Result.success(null))
        } else {
          callback(Result.success(null))
        }
      } catch (e: Exception) {
        callback(Result.success(null))
      }
    }
  }

  override fun getPendingUploadIds(callback: (Result<List<String>>) -> Unit) {
    callback(Result.success(activeUploads.keys.toList()))
  }

  private fun buildMetadata(data: TusUploadData): Map<String, String> {
    // Note: Do NOT Base64 encode values here - tus-java-client handles encoding automatically
    val metadata = mutableMapOf<String, String>()
    metadata["filename"] = data.filename
    metadata["deviceAssetId"] = data.deviceAssetId
    metadata["deviceId"] = data.deviceId
    metadata["fileCreatedAt"] = data.fileCreatedAt
    metadata["fileModifiedAt"] = data.fileModifiedAt
    metadata["isFavorite"] = data.isFavorite.toString()

    data.duration?.let { metadata["duration"] = it }
    data.livePhotoVideoId?.let { metadata["livePhotoVideoId"] = it }

    return metadata
  }

  private fun fetchAssetIdFromServer(
    uploadUrl: URL?,
    headers: Map<String, String>
  ): String? {
    if (uploadUrl == null) return null
    return try {
      val connection = uploadUrl.openConnection() as java.net.HttpURLConnection
      connection.requestMethod = "HEAD"
      connection.setRequestProperty("Tus-Resumable", "1.0.0")
      headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }
      connection.connect()

      val assetId = connection.getHeaderField("X-Immich-Asset-Id")
        ?: connection.getHeaderField("x-immich-asset-id")

      connection.disconnect()
      assetId
    } catch (e: Exception) {
      null
    }
  }
}
